File: acs-integration-tests/src/test/java/com/ge/predix/test/utils/ACSITSetUpFactoryPublic.java
/*******************************************************************************
 * Copyright 2017 General Electric Company
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 *******************************************************************************/

package com.ge.predix.test.utils;

import java.io.IOException;
import java.util.Arrays;
import java.util.Collections;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Scope;
import org.springframework.http.HttpHeaders;
import org.springframework.security.oauth2.client.OAuth2RestTemplate;
import org.springframework.stereotype.Component;

import com.ge.predix.acs.rest.Zone;

@Component
@Scope("prototype")
public class ACSITSetUpFactoryPublic implements ACSITSetUpFactory {

    private String acsZone1Name;
    private String acsZone2Name;
    private String acsZone3Name;

    @Value("${ACS_TESTING_UAA}")
    private String uaaUrl;

    @Value("${UAA_ADMIN_SECRET:adminsecret}")
    private String uaaAdminSecret;

    @Value("${ACS_SERVICE_ID:predix-acs}")
    private String serviceId;

    private static final String OAUTH_ENDPOINT = "/oauth/token";

    private String acsUrl;
    private HttpHeaders zone1Headers;
    private HttpHeaders zone3Headers;
    private OAuth2RestTemplate acsAdminRestTemplate;
    private OAuth2RestTemplate acsZonesAdminRestTemplate;
    private OAuth2RestTemplate acsZone1RestTemplate;
    private OAuth2RestTemplate acsZone2RestTemplate;
    private OAuth2RestTemplate acsReadOnlyRestTemplate;
    private OAuth2RestTemplate acsNoPolicyScopeRestTemplate;
    private Zone zone1;
    private Zone zone2;

    private UAAACSClientsUtil uaaTestUtil;

    @Autowired
    private ZoneFactory zoneFactory;

    @Override
    public void setUp() throws IOException {
        // TestConfig.setupForEclipse(); // Starts ACS when running the test in eclipse.
        this.acsUrl = this.zoneFactory.getAcsBaseURL();

        this.acsZone1Name = ZoneFactory.getRandomName(this.getClass().getSimpleName());
        this.acsZone2Name = ZoneFactory.getRandomName(this.getClass().getSimpleName());
        this.acsZone3Name = ZoneFactory.getRandomName(this.getClass().getSimpleName());

        this.zone1Headers = ACSTestUtil.httpHeaders();
        this.zone1Headers.set(PolicyHelper.PREDIX_ZONE_ID, this.acsZone1Name);

        this.zone3Headers = ACSTestUtil.httpHeaders();
        this.zone3Headers.set(PolicyHelper.PREDIX_ZONE_ID, this.acsZone3Name);

        this.uaaTestUtil = new UAAACSClientsUtil(this.uaaUrl, this.uaaAdminSecret);

        this.acsAdminRestTemplate = this.uaaTestUtil.createAcsAdminClientAndGetTemplate(this.acsZone1Name);
        this.acsZonesAdminRestTemplate = this.uaaTestUtil
                .createAcsAdminClient(Arrays.asList(this.acsZone1Name, this.acsZone2Name, this.acsZone3Name));
        this.acsReadOnlyRestTemplate = this.uaaTestUtil.createReadOnlyPolicyScopeClient(this.acsZone1Name);
        this.acsNoPolicyScopeRestTemplate = this.uaaTestUtil.createNoPolicyScopeClient(this.acsZone1Name);
        this.zone1 = this.zoneFactory.createTestZone(this.acsAdminRestTemplate, this.acsZone1Name,
                Collections.singletonList(this.uaaUrl + OAUTH_ENDPOINT));
        this.acsZone1RestTemplate = this.uaaTestUtil.createZoneClientAndGetTemplate(this.acsZone1Name, this.serviceId);

        this.zone2 = this.zoneFactory.createTestZone(this.acsAdminRestTemplate, this.acsZone2Name,
                Collections.singletonList(this.uaaUrl + OAUTH_ENDPOINT));
        this.acsZone2RestTemplate = this.uaaTestUtil.createZoneClientAndGetTemplate(this.acsZone2Name, this.serviceId);
    }

    @Override
    public void destroy() {
        this.zoneFactory.deleteZone(this.acsAdminRestTemplate, this.acsZone1Name);
        this.zoneFactory.deleteZone(this.acsAdminRestTemplate, this.acsZone2Name);
        this.uaaTestUtil.deleteClient(this.acsAdminRestTemplate.getResource().getClientId());
        this.uaaTestUtil.deleteClient(this.acsZone1RestTemplate.getResource().getClientId());
        this.uaaTestUtil.deleteClient(this.acsZone2RestTemplate.getResource().getClientId());
        this.uaaTestUtil.deleteClient(this.acsReadOnlyRestTemplate.getResource().getClientId());
        this.uaaTestUtil.deleteClient(this.acsNoPolicyScopeRestTemplate.getResource().getClientId());
        this.uaaTestUtil.deleteClient(this.acsZonesAdminRestTemplate.getResource().getClientId());
    }

    @Override
    public String getAcsUrl() {

        return this.acsUrl;
    }

    @Override
    public HttpHeaders getZone1Headers() {

        return this.zone1Headers;
    }

    @Override
    public OAuth2RestTemplate getAcsZoneAdminRestTemplate() {
        return this.acsZone1RestTemplate;
    }

    @Override
    public OAuth2RestTemplate getAcsZone2AdminRestTemplate() {

        return this.acsZone2RestTemplate;
    }

    @Override
    public OAuth2RestTemplate getAcsReadOnlyRestTemplate() {

        return this.acsReadOnlyRestTemplate;
    }

    @Override
    public OAuth2RestTemplate getAcsNoPolicyScopeRestTemplate() {

        return this.acsNoPolicyScopeRestTemplate;
    }

    @Override
    public Zone getZone1() {
        return this.zone1;
    }

    @Override
    public Zone getZone2() {
        return this.zone2;
    }

    @Override
    public String getAcsZone1Name() {
        return this.acsZone1Name;
    }

    @Override
    public String getAcsZone2Name() {
        return this.acsZone2Name;
    }

    @Override
    public String getAcsZone3Name() {
        return this.acsZone3Name;
    }

    @Override
    public HttpHeaders getZone3Headers() {
        return this.zone3Headers;
    }

    @Override
    public OAuth2RestTemplate getAcsZonesAdminRestTemplate() {
        return this.acsZonesAdminRestTemplate;
    }

    @Override
    public OAuth2RestTemplate getAcsZoneConnectorAdminRestTemplate(final String zone) {
        return this.uaaTestUtil.createAdminConnectorScopeClient(zone);
    }

    @Override
    public OAuth2RestTemplate getAcsZoneConnectorReadRestTemplate(final String zone) {
        return this.uaaTestUtil.createReadOnlyConnectorScopeClient(zone);
    }

    @Override
    public OAuth2RestTemplate getAcsAdminRestTemplate(final String zone) {
        return new UAAACSClientsUtil(this.uaaUrl, this.uaaAdminSecret).createAcsAdminClient(Arrays.asList(zone));
    }
}

File: acs-integration-tests/src/test/kotlin/org/eclipse/keti/test/utils/ACSITSetUpFactoryPublic.kt
/*******************************************************************************
 * Copyright 2018 General Electric Company
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package org.eclipse.keti.test.utils

import org.eclipse.keti.acs.rest.Zone
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Scope
import org.springframework.http.HttpHeaders
import org.springframework.security.oauth2.client.OAuth2RestTemplate
import org.springframework.stereotype.Component
import java.io.IOException

private const val OAUTH_ENDPOINT = "/oauth/token"

@Component
@Scope("prototype")
class ACSITSetUpFactoryPublic : ACSITSetUpFactory {

    override lateinit var acsUrl: String

    override lateinit var zone1: Zone

    override lateinit var zone2: Zone

    override lateinit var acsZone1Name: String

    override lateinit var acsZone2Name: String

    override lateinit var acsZone3Name: String

    override lateinit var zone1Headers: HttpHeaders

    override lateinit var zone3Headers: HttpHeaders

    override lateinit var acsZoneAdminRestTemplate: OAuth2RestTemplate

    override lateinit var acsZone2AdminRestTemplate: OAuth2RestTemplate

    override lateinit var acsReadOnlyRestTemplate: OAuth2RestTemplate

    override lateinit var acsNoPolicyScopeRestTemplate: OAuth2RestTemplate

    override lateinit var acsZonesAdminRestTemplate: OAuth2RestTemplate

    @Value("\${ACS_TESTING_UAA}")
    private lateinit var uaaUrl: String

    @Value("\${UAA_ADMIN_SECRET:adminsecret}")
    private lateinit var uaaAdminSecret: String

    @Value("\${ACS_SERVICE_ID:predix-acs}")
    private lateinit var serviceId: String

    private var acsAdminRestTemplate: OAuth2RestTemplate? = null

    private var uaaTestUtil: UaaAcsClientsUtil? = null

    @Autowired
    private lateinit var zoneFactory: ZoneFactory

    @Throws(IOException::class)
    override fun setUp() {
        // TestConfig.setupForEclipse(); // Starts ACS when running the test in eclipse.
        this.acsUrl = this.zoneFactory.acsBaseURL

        this.acsZone1Name = getRandomName(this.javaClass.simpleName)
        this.acsZone2Name = getRandomName(this.javaClass.simpleName)
        this.acsZone3Name = getRandomName(this.javaClass.simpleName)

        this.zone1Headers = httpHeaders()
        this.zone1Headers.set(PREDIX_ZONE_ID, this.acsZone1Name)

        this.zone3Headers = httpHeaders()
        this.zone3Headers.set(PREDIX_ZONE_ID, this.acsZone3Name)

        this.uaaTestUtil = UaaAcsClientsUtil(this.uaaUrl, this.uaaAdminSecret)

        this.acsAdminRestTemplate = this.uaaTestUtil!!.createAcsAdminClientAndGetTemplate(this.acsZone1Name)
        this.acsZonesAdminRestTemplate = this.uaaTestUtil!!
            .createAcsAdminClient(listOf<String>(this.acsZone1Name, this.acsZone2Name, this.acsZone3Name))
        this.acsReadOnlyRestTemplate = this.uaaTestUtil!!.createReadOnlyPolicyScopeClient(this.acsZone1Name)
        this.acsNoPolicyScopeRestTemplate = this.uaaTestUtil!!.createNoPolicyScopeClient(this.acsZone1Name)
        this.zone1 = this.zoneFactory.createTestZone(
            this.acsAdminRestTemplate!!, this.acsZone1Name,
            listOf(this.uaaUrl + OAUTH_ENDPOINT)
        )
        this.acsZoneAdminRestTemplate =
            this.uaaTestUtil!!.createZoneClientAndGetTemplate(this.acsZone1Name, this.serviceId)

        this.zone2 = this.zoneFactory.createTestZone(
            this.acsAdminRestTemplate!!, this.acsZone2Name,
            listOf(this.uaaUrl + OAUTH_ENDPOINT)
        )
        this.acsZone2AdminRestTemplate =
            this.uaaTestUtil!!.createZoneClientAndGetTemplate(this.acsZone2Name, this.serviceId)
    }

    override fun destroy() {
        this.zoneFactory.deleteZone(this.acsAdminRestTemplate!!, this.acsZone1Name)
        this.zoneFactory.deleteZone(this.acsAdminRestTemplate!!, this.acsZone2Name)
        this.uaaTestUtil!!.deleteClient(this.acsAdminRestTemplate!!.resource.clientId)
        this.uaaTestUtil!!.deleteClient(this.acsZoneAdminRestTemplate.resource.clientId)
        this.uaaTestUtil!!.deleteClient(this.acsZone2AdminRestTemplate.resource.clientId)
        this.uaaTestUtil!!.deleteClient(this.acsReadOnlyRestTemplate.resource.clientId)
        this.uaaTestUtil!!.deleteClient(this.acsNoPolicyScopeRestTemplate.resource.clientId)
        this.uaaTestUtil!!.deleteClient(this.acsZonesAdminRestTemplate.resource.clientId)
    }

    override fun getAcsZoneConnectorAdminRestTemplate(zone: String): OAuth2RestTemplate {
        return this.uaaTestUtil!!.createAdminConnectorScopeClient(zone)
    }

    override fun getAcsZoneConnectorReadRestTemplate(zone: String): OAuth2RestTemplate {
        return this.uaaTestUtil!!.createReadOnlyConnectorScopeClient(zone)
    }

    override fun getAcsAdminRestTemplate(zone: String): OAuth2RestTemplate {
        return UaaAcsClientsUtil(this.uaaUrl, this.uaaAdminSecret).createAcsAdminClient(listOf(zone))
    }
}
